design/commands.md

```diff
@@ -0,0 +1,201 @@
+# Commands (Sharing Binaries)
+
+## Context
+
+Currently, there's no clear way to share a binaries (or scripts) across multiple jobs.  Some people have chosen to share via git repositories, yum packages, node modules, and even docker images.  But nothing is standardized and works easily in all containers.
+
+## Status
+
+- 11 September 2017: Proposal submitted
+
+## Proposal
+
+Have a single interface for executing a versioned command (via remote binary, docker image, or habitat package) during a Screwdriver build.  Managed and discoverable through the Screwdriver interface.
+
+This provides a single abstraction for execution of shared code.
+
+```bash
+$ sd_cmd chefdk/knife@12 search node '*:*'
+$ sd_cmd nodejs/vuln_check@2 ./package.json
+```
+
+## Details
+
+### Specification
+
+```yaml
+# Namespace for the command
+namespace: foo
+# Command name itself
+command: bar
+# Description of the command and what it does
+description: |
+  Lorem ipsum dolor sit amet.
+# Major and Minor version number (patch is automatic)
+version: 1.0
+# Format the command is in (see below)
+format: habitat
+
+# Habitat config
+habitat:
+    mode: remote
+    package: core/git/2.14.1
+    # If local
+    # mode: local
+    # package: ./foobar.hart
+    binary:  git
+
+# Docker config
+docker:
+    image: node:1.2.3
+
+# Binary config
+binary:
+    file: ./foobar.sh
+```
+
+### Versioning
+
+Published command versions must be following [semver](http://semver.org)
+
+    [MAJOR].[MINOR].[PATCH]
+
+This is to make it easy for people to get updates without the risk of breaking backwards compatibility.
+
+Requesting published versions follows the semver resolution as well:
+
+ - **X-Ranges:** `1.2.*` `1.2` `1.x` `1` `*`
+
+    Any of x, *, or blank may be used to "stand in" for one of the numeric values in the [major, minor, patch] tuple.
+
+ - **Tilde Ranges:** `~1.2.3` `~1.2` `~1`
+
+    Allows patch-level changes if a minor version is specified on the comparator. Allows minor-level changes if not.
+
+ - **Caret Ranges:** `^1.2.3` `^0.2.5` `^0.0.4`
+
+    Allows changes that do not modify the left-most non-zero digit in the [major, minor, patch] tuple. In other words, this allows patch and minor updates for versions 1.0.0 and above, patch updates for versions 0.X >=0.1.0, and no updates for versions 0.0.X.
+
+ - **Explicit Pinning:** `1.2.3` `1.5.3`
+
+    Allows only that exact version number, nothing else.
+
+ - **Tags:** `latest` `stable` `feature-abc`
```

Comment:
This needs to be alphanumeric. Otherwise, if tag is `123`, we won't know if it's referring to tag 123 or version 123?

Reply:
Hmm, `npm` checks if the tag is valid semver.  If it is, it rejects it: https://docs.npmjs.com/cli/dist-tag#caveats

We can either follow that or require that it starts with a letter and only matches `/[a-z][a-z0-9-]+/`

Reply:
Yeah I think we should check, similar to how template tags are.